### PR TITLE
DEV: Optimizer Engine classes

### PR DESCRIPTION
**Pull request by Corwinpro**
This PR implements base optimization engine class `BaseOptimizerEngine`, and one optimizer: `WeightedOptimizer`.

### Summary

Optimizer Engine performs black-box optimization (minimization) of the KPI values by exploring the space of input parameters. The way the input space is explored depends on the `OptimizerEngine` implementation. Users should be able to choose an Optimizer Engine via `MCOModel`, setup the optimization parameters, and get optimized data from the OptimizerEngine.optimize method.

The `OptimizerEngine` is an backend tool, available to developers of use cases out of the box as a part of the `MCO`. We will introduce `WeightedOptimizer` to the BDSS core. 

Instantiation of an `OptimizerEngine` is delegated to the `MCOModel`. Each `MCO` is supposed to manage one optimizer engine.

**Additional dependencies**

One additional library is required to run the optimization process: `scipy>=1.2.1`

### General API

The setup of the optimization engine:
```
optimizer = AnOptimizerEngine(
    kpis=model.kpis,
    parameters=model.parameters,
    single_point_evaluator=evaluator,
    verbose_run=model.verbose_run,
    **kwargs,
)
```

The main public method is `optimize`:

```
    def optimize(self):
        """ Main entry point to the OptimizerEngine. This is an general
        iterator over [explored input space, objective space, **kwargs].
        It can yield data as a single workflow evaluation has been completed,
        or return the optimization data after all required computations
        have finished.
        """
```

### Done
- Implemented `BaseOptimizerEngine`: it introduces the abstract `optimize` method, and a few useful utilities (converting mixed minimization-maximization problem into minimization, wrapper function for `evaluator.evaluate()` API)
- Added `WeightedOptimizer` from the ITWM example plugin, with small refactoring of utility methods
- `RangedMCOParameter`: new `initial_value` attribute added. This attribute allows the user to specify a prior estimate
- Added `MCOParameter` and `OptimizerEngine` references to the `force_bdss.api`
- `NevergradOptimizer` now becomes a part of a stand-alone plugin [force-nevergrad](https://github.com/force-h2020/force-bdss-plugin-nevergrad)

### CODE REVIEW

**Comment by Corwinpro** [comment]
I believe there is a strong need in refactoring of `KPISpecification`, or implementing engine-specific subtypes. For example, `nevergrad` requires `upper_bound` on KPI: currently, we do hacky thing and use `scale_factor` instead. This is very bad, and is obviously not clear to the user how to specify the upper bound explicitly.

**Comment by flongford** [comment]
Note: a new repo [`force-bdss-plugin-nevergrad`](https://github.com/force-h2020/force-bdss-plugin-nevergrad) has been created to maintain support for the `nevergrad` library

File: ci/__main__.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 PIP_DEPS = ["stevedore==1.30.1"]
 
-ADDITIONAL_CORE_DEPS = ["numpy==1.15.4-2"]
+ADDITIONAL_CORE_DEPS = ["numpy==1.15.4-2", "scipy>=1.2.1"]
```
[the remark this answers is not in the document]
Comment (flongford) [reply ci/__main__.py]: No, I see its due to `WeightedOptimizerEngine`

File: force_bdss/mco/optimizer_engines/base_optimizer_engine.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import abc
+import logging
+
+from traits.api import ABCHasStrictTraits, List, Instance, Bool
+
+from force_bdss.api import KPISpecification, BaseMCOParameter, IEvaluator
```
[the remark this answers is not in the document]
Comment (Corwinpro) [reply force_bdss/mco/optimizer_engines/base_optimizer_engine.py]: Makes perfect sense to me. Will do, thank you.

**Pull request merged**